jupyter-extension/jupyterlab_pachyderm/tests/test_integrations.py

```diff
@@ -540,11 +541,65 @@ def test_pps(dev_server, simple_pachyderm_env, notebook_path):
     )
 
 
-def test_pps_validation_errors(dev_server, notebook_path):
+def test_pps_last_modified_time_not_specified_validation(dev_server, notebook_path):
     r = requests.put(f"{BASE_URL}/pps/_create/{notebook_path}", data=json.dumps({}))
     assert r.status_code == 400, r.text
     assert r.json()["reason"] == f"Bad Request: last_modified_time not specified"
 
+@pytest.mark.parametrize("simple_pachyderm_env", [True], indirect=True)
+def test_pps_external_files_do_not_exist_validation(
+    dev_server, simple_pachyderm_env, notebook_path
+):
+    _client, repo, _companion_repo, pipeline = simple_pachyderm_env
+
+    new_notebook_data = _update_metadata(TEST_NOTEBOOK, repo, pipeline, 'does_not_exist.py')
+    notebook_path.write_text(new_notebook_data)
+    last_modified = datetime.utcfromtimestamp(os.path.getmtime(notebook_path))
+    data = dict(last_modified_time=f"{datetime.isoformat(last_modified)}Z")
+    r = requests.put(
+        f"{BASE_URL}/pps/_create/{notebook_path}", data=json.dumps(data)
+    )
+    
+    assert r.status_code == 400
+    assert r.json()["message"] == 'Bad Request'
+    assert r.json()["reason"] == 'external file does_not_exist.py could not be found in the directory of the Jupyter notebook'
+
+@pytest.mark.parametrize("simple_pachyderm_env", [True], indirect=True)
+def test_pps_uploads_external_files(
+    dev_server, simple_pachyderm_env, notebook_path
+):
+    client, repo, companion_repo, pipeline = simple_pachyderm_env
+    new_notebook_data = _update_metadata(TEST_NOTEBOOK, repo, pipeline, 'hello.py,world.py')
+    notebook_path.write_text(new_notebook_data)
+    last_modified = datetime.utcfromtimestamp(os.path.getmtime(notebook_path))
+    data = dict(last_modified_time=f"{datetime.isoformat(last_modified)}Z")
+    try:
+        TEST_NOTEBOOK.with_name("hello.py").write_text("print('hello')")
+        TEST_NOTEBOOK.with_name("world.py").write_text("print('world')")
+
+        r = requests.put(
+            f"{BASE_URL}/pps/_create/{notebook_path}", data=json.dumps(data)
+        )
+
+        assert r.status_code == 200, r.text
+        job_info = next(client.pps.list_job(pipeline=pipeline))
+        job_info = client.pps.inspect_job(job=job_info.job, wait=True)
+        assert job_info.state == pps.JobState.JOB_SUCCESS
+        assert r.json()["message"] == (
+            "Create pipeline request sent. You may monitor its "
+            'status by running "pachctl list pipelines" in a terminal.'
+        )
+        commits = [info.commit.id for info in client.pfs.list_commit(repo=companion_repo)]
+        assert len(commits) == 1
+        file_uri = f'{companion_repo.project.name}/{companion_repo.name}@{commits[0]}:'
```

Comment:
Ah so the project is optional? I can update this it's a quick change 👍🏻

Reply:
I think this is necessary because the tests create a new test project on each run. Without it default is used and it will throw a file not found error.

```
ERROR    root:_channel.py:246 Exception in callback <function MetadataClientInterceptor.intercept.<locals>.<lambda> at 0x7f178cb6ae50>: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.NOT_FOUND
        details = "commit default/test_pipeline_602794__context@6d9285851e964e0596a9847c5a05527c not found"
        debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2024-01-10T00:47:48.608332695+00:00", grpc_status:5, grpc_message:"commit default/test_pipeline_602794__context@6d9285851e964e0596a9847c5a05527c not found"}"
```

Reply:
I believe that the companion repo is set with the correct project. I just meant that the `pfs.Repo` object should have a string representation that is equivalent to `f"{repo.project.name}/{repo.name}"`

Reply:
Ahh that makes sense I will adjust this to take advantage of the `pfs.Repo` string representation.